microsoft/testsuites/xdp/xdptools.py

```diff
@@ -0,0 +1,134 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+import re
+from pathlib import PurePath
+from typing import Any, Dict
+
+from lisa import LisaException, UnsupportedDistroException
+from lisa.executable import Tool
+from lisa.operating_system import Debian, Fedora
+from lisa.tools import Ethtool, Git, Make
+from lisa.tools.ethtool import DeviceGroLroSettings
+from lisa.util import find_groups_in_lines
+
+
+class XdpTool(Tool):
+    """
+    The community xdp tools, it's used to verify XDP by community test cases.
+    """
+
+    _xdp_tools_repo = "https://github.com/xdp-project/xdp-tools.git"
+    _xdp_tools_tag = "v1.2.0"
+    _default_command = "xdptool"
+
+    #     [test_ether_deny]             PASS
+    #     [test_python_basic]           SKIPPED
+    _xdp_test_result_pattern = re.compile(
+        r"^\s+\[(?P<name>\w+)]\s+(?P<result>\w+)\r?$", re.M
+    )
+
+    @property
+    def command(self) -> str:
+        return str(self._command)
+
+    @property
+    def can_install(self) -> bool:
+        ethtool = self.node.tools[Ethtool]
+        statistics = ethtool.get_device_statistics(self.node.nics.default_nic)
+
+        # check if xdp supported on nic
+        if not any("xdp_drop" in x for x in statistics):
+            raise UnsupportedDistroException(
+                self.node.os,
+                "Cannot find xdp_drop in ethotool statistics. "
+                "It means it doesn't support XDP.",
+            )
+
+        return True
+
+    def run_full_test(self) -> None:
+        """
+        run full test of xdp tools repo
+        """
+        result = self.node.execute("make test", sudo=True, cwd=self._code_path)
+
+        abnormal_results: Dict[str, str] = {}
+        for item in find_groups_in_lines(
+            result.stdout, pattern=self._xdp_test_result_pattern
+        ):
+            if item["result"] not in ["PASS", "SKIPPED"]:
+                abnormal_results[item["name"]] = item["result"]
+        if abnormal_results:
+            raise LisaException(f"found failed tests: {abnormal_results}")
+        result.assert_exit_code(
+            0, "unknown error on xdp tests, please check log for more details."
+        )
+
+    def _initialize(self, *args: Any, **kwargs: Any) -> None:
+        super()._initialize(*args, **kwargs)
+        self._command: PurePath = PurePath(self._default_command)
+        self._gro_lro_settings: Dict[str, DeviceGroLroSettings] = {}
+
+    def _install(self) -> bool:
+        # install dependencies
+
+        config_envs: Dict[str, str] = {}
+        if isinstance(self.node.os, Debian):
+            if self.node.os.information.version < "18.4.0":
```

Comment:
this works for ubuntu, how about debian images, current the latest debian image is 11

Reply:
this place is to compat with 1804. Lower version doesn't support xdp, higher version use the default toolchain. I haven't test with debian. Will fix debian problems after test with chaos monkey.